spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java

```diff
@@ -378,4 +506,84 @@ private String getStackTraceAsString(Throwable cause) {
 		return stringWriter.getBuffer().toString();
 	}
 
+	/**
+	 * Override this if you want different header names to be used
+	 * in the sent record.
+	 * @return the default Dlt header names
+	 */
+	protected HeaderNames getHeaderNames() {
+		return DeadLetterPublishingHeaderNamesBuilder
+				.original()
+					.offsetHeader(KafkaHeaders.DLT_ORIGINAL_OFFSET)
+					.timestampHeader(KafkaHeaders.DLT_ORIGINAL_TIMESTAMP)
+					.timestampTypeHeader(KafkaHeaders.DLT_ORIGINAL_TIMESTAMP_TYPE)
+					.topicHeader(KafkaHeaders.DLT_ORIGINAL_TOPIC)
+					.partitionHeader(KafkaHeaders.DLT_ORIGINAL_PARTITION)
+				.exception()
+					.keyExceptionFqcn(KafkaHeaders.DLT_KEY_EXCEPTION_FQCN)
+					.exceptionFqcn(KafkaHeaders.DLT_EXCEPTION_FQCN)
+					.keyExceptionMessage(KafkaHeaders.DLT_KEY_EXCEPTION_MESSAGE)
+					.exceptionMessage(KafkaHeaders.DLT_EXCEPTION_MESSAGE)
+					.keyExceptionStacktrace(KafkaHeaders.DLT_KEY_EXCEPTION_STACKTRACE)
+					.exceptionStacktrace(KafkaHeaders.DLT_EXCEPTION_STACKTRACE)
+				.build();
+	}
+
+	/**
+	 * Container class for the headers that will be added to the produced record.
+	 */
+	public static class HeaderNames {
+
+		private final HeaderNames.Original original;
+		private final HeaderNames.Exception exception;
+
+		HeaderNames(HeaderNames.Original original, HeaderNames.Exception exception) {
+			this.original = original;
+			this.exception = exception;
+		}
+
+		static class Original {
```

Comment:
Why do we have these classes here and in the builder as well?

Reply:
It's basically to keep the properties more organised, instead of a single class with 11 properties, and also provide a more organised API. But maybe I can make it simpler if you prefer.

Reply:
I understand original Vs. exception, my concern is duplication of similar classes in the builder.